airflow/api/common/experimental/trigger_dag.py

```diff
@@ -57,6 +58,12 @@ def _trigger_dag(
             dag_id
         ))
 
+    if UniqueExecutionDate and dag_run.find(dag_id=dag_id, execution_date=execution_date):
+        raise ExecutionDateAlreadyExists("Execution date {} already exists for dag id {}".format(
+            execution_date,
+            dag_id
+        ))
```

Comment:
This sort of pattern (of check for unique then insert) is prone to race conditions and isn't perfect. The only sure fire way to guarantee uniqueness is to have a DB constraint and handle the constraint violation that the DB will raise.

Reply:
Currently the primary key of dag_run table is integer “id”, not DAG I’d + execution_date.

Reply:
But I agree changing from DB constraint level is a better solution.

Reply:
Sorry I checked again. Yes there is already an Unique constraint for dag_id + execition_date.

Reply:
in fact the original reason of the problem is the fact that execution_date is unique (report for here AIRFLOW-2319 for example).

The problem is when the sqlalchemy error is raised on API endpoint call: 
- error message is not understandable
- internal server error happen (and not a correct error message)

I must admit that with my time we have spend days trying to find the reason of that error.

If we keeping the unique execution_date, It could be nice for future user to have a error message more easy to understand :-) .

What are you thinking ?